src/main/scala/org/infinispan/spark/rdd/InfinispanRDD.scala

```diff
@@ -20,30 +22,37 @@ class InfinispanRDD[K, V](@transient val sc: SparkContext,
 
-   @transient lazy val remoteCache: RemoteCache[K,V] = {
-      val remoteCacheManager = new RemoteCacheManager(new ConfigurationBuilder().withProperties(configuration).pingOnStartup(true).build())
+   private[rdd] def createBuilder(address: InetSocketAddress) = {
+      new ConfigurationBuilder().withProperties(configuration)
+              .addServer().host(address.getHostString).port(address.getPort)
+              .balancingStrategy(new PreferredServerBalancingStrategy(address))
+   }
+
+   @transient lazy val remoteCache: RemoteCache[K, V] = {
+      val config = new ConfigurationBuilder().withProperties(configuration).pingOnStartup(true).build()
+      val remoteCacheManager = new RemoteCacheManager(config)
       sc.addSparkListener(new SparkListener {
          override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = remoteCache.stop()
       })
       val optCacheName = Option(configuration.getProperty(InfinispanRDD.CacheName))
-      optCacheName.map(name => remoteCacheManager.getCache[K,V](name)).getOrElse(remoteCacheManager.getCache[K,V])
+      optCacheName.map(name => remoteCacheManager.getCache[K,V](name)).getOrElse(remoteCacheManager.getCache[K, V])
    }
 
-   override def compute(split: Partition, context: TaskContext): Iterator[(K, V)] = {
+   def compute[R](split: Partition, context: TaskContext,
+               cacheManagerProvider: InetSocketAddress => RemoteCacheManager,
+               filterFactory: String, factoryParams: Array[AnyRef]): Iterator[(K, R)] = {
+
       logInfo(s"Computing partition $split")
       val infinispanPartition = split.asInstanceOf[InfinispanPartition]
       val config = infinispanPartition.properties
       val cacheName = config.getProperty(InfinispanRDD.CacheName)
       val batch = config.readWithDefault[Integer](InfinispanRDD.ReadBatchSize)(default = InfinispanRDD.DefaultReadBatchSize)
       val address = infinispanPartition.location.address.asInstanceOf[InetSocketAddress]
-      val builder = new ConfigurationBuilder().withProperties(configuration)
-              .addServer().host(address.getHostString).port(address.getPort)
-              .balancingStrategy(new PreferredServerBalancingStrategy(address))
-      val remoteCacheManager = new RemoteCacheManager(builder.build())
+      val remoteCacheManager = cacheManagerProvider(address)
       val cache = Option(cacheName).map(name => remoteCacheManager.getCache(name)).getOrElse(remoteCacheManager.getCache)
       val segmentFilter = infinispanPartition.segments.map(setAsJavaSet).orNull
-      val filterFactory = config.read[String](InfinispanRDD.FilterFactory)
-      val closeableIterator = cache.retrieveEntries(filterFactory.getOrElse(null.asInstanceOf[String]), segmentFilter, batch)
+      val closeableIterator = cache.retrieveEntries(filterFactory, factoryParams, segmentFilter, batch)
+
       context.addTaskCompletionListener(t => {
          closeableIterator.close()
          remoteCacheManager.stop()
```

Comment:
Is this where you make sure remote cache manager gets stopped, right? Now that I see where you stop the remote cache manager, how often are you going to be creating/stopping remote cache managers? Wondering if some caching could be done? Again, connection pools start/stop, thread pools...etc.